atc/db/build_in_memory_check.go

```diff
@@ -0,0 +1,560 @@
+func (b *inMemoryCheckBuild) Name() string                            { return CheckBuildName }
+func (b *inMemoryCheckBuild) TeamID() int                             { return b.checkable.TeamID() }
+func (b *inMemoryCheckBuild) TeamName() string                        { return b.checkable.TeamName() }
+func (b *inMemoryCheckBuild) PipelineID() int                         { return b.checkable.PipelineID() }
+func (b *inMemoryCheckBuild) PipelineName() string                    { return b.checkable.PipelineName() }
+func (b *inMemoryCheckBuild) PipelineRef() atc.PipelineRef            { return b.checkable.PipelineRef() }
+func (b *inMemoryCheckBuild) Pipeline() (Pipeline, bool, error)       { return b.checkable.Pipeline() }
+func (b *inMemoryCheckBuild) ResourceID() int                         { return b.resourceId }
+func (b *inMemoryCheckBuild) ResourceName() string                    { return b.resourceName }
+func (b *inMemoryCheckBuild) ResourceTypeID() int                     { return b.resourceTypeId }
+func (b *inMemoryCheckBuild) Schema() string                          { return schema }
+func (b *inMemoryCheckBuild) IsRunning() bool                         { return b.running }
+func (b *inMemoryCheckBuild) IsManuallyTriggered() bool               { return false }
+func (b *inMemoryCheckBuild) SpanContext() propagation.TextMapCarrier { return b.spanContext }
+func (b *inMemoryCheckBuild) PipelineInstanceVars() atc.InstanceVars {
+	return b.checkable.PipelineInstanceVars()
+}
+
+// JobID returns 0 because check build doesn't belong to any job.
+func (b *inMemoryCheckBuild) JobID() int { return 0 }
+
+// JobName returns an empty string because check build doesn't belong to any job.
+func (b *inMemoryCheckBuild) JobName() string { return "" }
+
+func (b *inMemoryCheckBuild) LagerData() lager.Data {
+	data := lager.Data{
+		"team":     b.TeamName(),
+		"pipeline": b.PipelineName(),
+	}
+
+	if b.preId != 0 {
+		data["preBuildId"] = b.preId
+	}
+
+	if b.id != 0 {
+		data["build"] = b.id
+	}
+
+	if b.resourceId != 0 {
+		data["resource"] = b.resourceName
+	}
+
+	if b.resourceTypeId != 0 {
+		data["resourceType"] = b.resourceTypeName
+	}
+
+	return data
+}
+
+func (b *inMemoryCheckBuild) TracingAttrs() tracing.Attrs {
+	attrs := tracing.Attrs{
+		"team":     b.TeamName(),
+		"pipeline": b.PipelineName(),
+	}
+
+	if b.preId != 0 {
+		attrs["preBuildId"] = fmt.Sprintf("%d", b.preId)
+	}
+
+	if b.id != 0 {
+		attrs["build"] = fmt.Sprintf("%d", b.id)
+	}
+
+	if b.resourceId != 0 {
+		attrs["resource"] = b.resourceName
+	}
+
+	if b.resourceTypeId != 0 {
+		attrs["resourceType"] = b.resourceTypeName
+	}
+
+	return attrs
+}
+
+// Reload just does nothing because an in-memory build lives shortly.
+func (b *inMemoryCheckBuild) Reload() (bool, error) {
+	return true, nil
+}
+
+func (b *inMemoryCheckBuild) PrivatePlan() atc.Plan {
+	if !b.running {
+		panic("not-implemented")
```

Comment:
For `panic`, let's discuss in your other comment that looks dedicate for `panic`.

Here, if `b.running` is false, private plan is unavailable. I think explicitly `panic` is better than returning an empty plan. Maybe I can revise the panic message?

Reply:
I think the point I was trying to make is that you are using panics for when the application is not in the right state, basically using it for assertions of the app being in a specific state. I just find it to be a pretty destructive way of catching something that went wrong because it will bring down the entire concourse, especially if this incorrect state will not cause any harm for the web or worker node to continue running. Thats why honestly I think best would be to make these impossible state unrepresentable, but if that is not an option then maybe we can trust that the application is in that correct state. Once you get to the part of the code that actually needs to do an assertion on the state (I guess in this case would be that the build is in a running state) we can error and log if it is not? This is all just my opinion though, so I'd like to know what you think

Reply:
Sorry, I have should resolved this comment. In the latest commit, I have updated the code to just return the plan.